nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/service/ColdStorageServiceImpl.java

```diff
@@ -54,84 +77,134 @@
+    public void registerContribution(Object contribution, String extensionPoint, ComponentInstance contributor) {
+        if (COLDSTORAGE_RENDITION_EP.equals(extensionPoint)) {
+            ColdStorageRenditionDescriptor descriptor = (ColdStorageRenditionDescriptor) contribution;
+            String docType = descriptor.getDocType();
+            if (docType != null) {
+                renditionByDocType.put(docType, descriptor.getRenditionName());
+            }
+            String facet = descriptor.getFacet();
+            if (facet != null) {
+                renditionByFacets.put(docType, descriptor.getRenditionName());
+            }
+            if (docType == null && facet == null) {
+                defaultRendition = descriptor.getRenditionName();
+            }
+        } else {
+            log.error("Unknown extension point " + extensionPoint);
+        }
+    }
 
-    public static final String COLD_STORAGE_CONTENT_AVAILABLE_UNTIL_MAIL_TEMPLATE_KEY = "coldStorageAvailableUntil";
+    @Override
+    public void unregisterContribution(Object contribution, String extensionPoint, ComponentInstance contributor) {
+    }
 
-    public static final String COLD_STORAGE_CONTENT_AVAILABLE_NOTIFICATION_NAME = "ColdStorageContentAvailable";
+    public String getRenditionName(DocumentModel doc) {
+        if (renditionByDocType.containsKey(doc.getType())) {
+            return renditionByDocType.get(doc.getType());
+        }
+        for (Map.Entry<String, String> entry : renditionByFacets.entrySet()) {
+            if (doc.hasFacet(entry.getKey())) {
+                return entry.getValue();
+            }
+        }
 
-    /** @since 10.10 **/
-    public static final String COLD_STORAGE_CONTENT_RESTORED_NOTIFICATION_NAME = "ColdStorageContentRestored";
+        if (defaultRendition == null) {
+            throw new NuxeoException("Please contribute a default rendition name");
+        }
+        return defaultRendition;
+    }
 
-    public static final String COLD_STORAGE_CONTENT_ARCHIVE_LOCATION_MAIL_TEMPLATE_KEY = "archiveLocation";
+    @Override
+    public Blob getRendition(DocumentModel doc, CoreSession session) {
+        // retrieve the thumbnail which will be used to replace the content, once the move done
+        Blob blob = Framework.getService(ThumbnailService.class).getThumbnail(doc, session);
+        String renditionName = getRenditionName(doc);
+        if (renditionName != null) {
+            try {
+                RenditionService renditionService = Framework.getService(RenditionService.class);
+                Rendition rendition = renditionService.getRendition(doc, renditionName);
+                blob = rendition.getBlob();
+            } catch (NuxeoException e) {
+                log.error("Cannot retrieve the rendition definition {}", e.getMessage());
+                throw new NuxeoException(String.format("Cannot retrieve the rendition for document %s.", doc),
+                        SC_NOT_FOUND);
```

Comment:
It is better to not log when you rethrow an exception as upper layers may log it too.

Furthermore, you could give the original exception to the one you're rethrowing.

Reply:
Hello,

It’s more clear now.
The original exception will be `SC_INTERNAL_SERVER_ERROR`. 
( is it enough clear ?, because this one is a generic error)

Reply:
In this case, it makes sense to change the status type:
```
throw new NuxeoException(String.format("Cannot retrieve the rendition for document %s.", doc), e, SC_NOT_FOUND);
```

Giving the exception will help us in production to identify the root cause if we need to analyze such errors.